<a href="https://colab.research.google.com/github/dibyarupnath/CodeClause_Plagiarism_Checker/blob/main/PlagiarismChecker_py.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Fetching required libraries

In [ ]:
import os
from numpy import vectorize 
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

In [ ]:
os.chdir("/content/drive/MyDrive/Projects/CodeClause AI Internship/Plagiarism Checker/")

## Loading the test files to check for plagiarism later

In [ ]:
sample_files = [doc for doc in os.listdir() if doc.endswith('.txt')]
sample_contents = [open(File).read() for File in sample_files]

## Vectorizing
+ Using lambda functions to convert text to array of numbers 
+ Vectorizing the text using Tfidf

In [ ]:
vectorize = lambda Text: TfidfVectorizer().fit_transform(Text).toarray()

In [ ]:
similarity = lambda doc1, doc2: cosine_similarity([doc1, doc2])

In [ ]:
vectors = vectorize(sample_contents)
s_vectors = list(zip(sample_files, vectors))

## Plagiarism Checker

In [ ]:
def check_plagiarism():
    results = set()
    global s_vectors
    for sample_a, text_vector_a in s_vectors:
        new_vectors = s_vectors.copy()
        current_index = new_vectors.index((sample_a, text_vector_a))
        del new_vectors[current_index]
        for sample_b, text_vector_b in new_vectors:
            sim_score = similarity(text_vector_a, text_vector_b)[0][1]
            sample_pair = sorted((sample_a, sample_b))
            score = sample_pair[0], sample_pair[1], sim_score
            results.add(score)
    return results

## Testing the Plagiarism Checker

In [ ]:
for data in check_plagiarism():
    print(data)

('artificial-intelligence.txt', 'environmental-pollution-2.txt', 0.23575336200723218)
('environmental-pollution-2.txt', 'internet.txt', 0.21772691345311806)
('artificial-intelligence.txt', 'environmental-pollution.txt', 0.2621776128782791)
('artificial-intelligence.txt', 'internet.txt', 0.2589825475776815)
('artificial-intelligence-2.txt', 'environmental-pollution-2.txt', 0.2844393682206794)
('artificial-intelligence-2.txt', 'internet.txt', 0.2645777745720631)
('artificial-intelligence-2.txt', 'internet-2.txt', 0.4275673317760499)
('artificial-intelligence-2.txt', 'artificial-intelligence.txt', 0.40973120483887754)
('environmental-pollution.txt', 'internet.txt', 0.41300636805603264)
('environmental-pollution-2.txt', 'internet-2.txt', 0.3069336737431845)
('environmental-pollution-2.txt', 'environmental-pollution.txt', 0.43936926094815654)
('artificial-intelligence-2.txt', 'environmental-pollution.txt', 0.30081344260255727)
('artificial-intelligence.txt', 'internet-2.txt', 0.313581526274

## Explanation 
I chose 3 topics for essays: environmental pollution, artificial intelligence and the internet. Then i googled 2 essays for each topic and pasted them into text files. 

We can see that unsurprisingly, the similarity between artificial-intelligence.txt and environmental-pollution-2.txt is only 0.2357 => 23.6%

and the similarity between artificial-intelligence-2.txt and artificial-intelligence.txt is a whopping 40.9%

and so on, taking a look at the amounts of similarity, we can understand that the plagiarism checker does a fair job at detecting similarities.